Feature/recover password

juanmaMacGyverCode opened the pull request:
This feature implements the recovery password when requested.

alu0100897562 committed:
[WIP] Adding AWS S3 service

ManuPerezDev committed:
[WIP] Changing proposal constrains and fixing multipart request

alu0100897562 committed:
[WIP] Fixing Multipart Handling on Proposal Controller

Nachocss committed:
Fixed: Multipart handler on ProposalApiController
Fixed: getBusinessDaysFrom method from ProposalDate

Nachocss committed:
Fixed: AmazonS3Client Bean is now been initialized correctly.

ManuPerezDev committed:
Fixed proposals validate and builder

Nachocss committed:
Uploaded files have now public access through their URL.
Fixed S3 tests

Nachocss committed:
[WIP] Configuring localstack's S3 endpoint and implementing Image storage e2e flow

ibaiul committed:
Improve performance of the file upload process and require AWS credentials to be provided through the ENV

Nachocss committed:
Refactor:
 * Created SkillDto and VolunteerDto and added to ProposalResponseDto
 * Changed AWS SES SDK version
 * Renamed columns on table proposals and added starting_proposal_date
 * Added skills and requirements to registerESALAndProposal Method of TestData

Nachocss committed:
Docker-compose now creates an S3 bucket

ManuPerezDev committed:
Implemented proposal creation endpoint as reviser
Added volunteer curriculum vitae storage

ManuPerezDev committed:
Fixed RemoteStorageService's extension check and assigment

ManuPerezDev committed:
Added validation for proposal closing deadline

ManuPerezDev committed:
Refactor proposal validate

ManuPerezDev committed:
Remove jenkins unsupported temporal units

ManuPerezDev committed:
Remove jenkins unsupported temporal units

ManuPerezDev committed:
Refactor proposal validate

Nachocss committed:
Published proposals can now be fetched as a paginated list, sorted by the proximity of their closing date

Nachocss committed:
Removed unnecessary data of the listed proposals, according to Product team requirements

Nachocss committed:
Revisers are now notified via email when a new proposal is created.
[WIP] Revisers can submit their review and send it to the ESAL's contact person

Nachocss committed:
Added test for revision submission of proposals

Nachocss committed:
Added name and surname as columns on credentials table and as fields on entity User.
Created entity Reviser.
Revision email submission is now null safe.
Refactor:
 * JpaCredential name changed.
 * Some valueobjects moved to entities package.

ManuPerezDev committed:
Add proposal revision template without feedback and minor refactor

Nachocss committed:
Add proposal status and removed published bool property

mariasoria committed:
Changes in JpaProposal and JpaStatusRepository. Fixed tests related to proposal status.

Nachocss committed:
Merge remote-tracking branch 'origin/develop' into feature/revision-required-notification-service

# Conflicts:
#	backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
#	backend/src/main/java/com/huellapositiva/application/dto/ProposalLiteDto.java
#	backend/src/main/java/com/huellapositiva/application/dto/ProposalRequestDto.java
#	backend/src/main/java/com/huellapositiva/application/dto/ProposalResponseDto.java
#	backend/src/main/java/com/huellapositiva/domain/actions/FetchPaginatedProposalsAction.java
#	backend/src/main/java/com/huellapositiva/domain/actions/FetchProposalAction.java
#	backend/src/main/java/com/huellapositiva/domain/model/entities/Proposal.java
#	backend/src/main/java/com/huellapositiva/domain/repository/ProposalRepository.java
#	backend/src/main/java/com/huellapositiva/infrastructure/orm/entities/JpaProposal.java
#	backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaProposalRepository.java
#	backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java
#	backend/src/test/java/com/huellapositiva/util/TestData.java

mariasoria committed:
Add comments and refactor

mariasoria committed:
Merge remote-tracking branch 'origin/feature/revision-required-notification-service' into feature/revision-required-notification-service

Nachocss committed:
Fixed/Contact Person email confirmation now set the correct role

Nachocss committed:
Refactor/Removed "Express" value objects

Nachocss committed:
[WIP] Fetching unavailable proposals redirects to proposal listing page. Entities Proposal status can now be FINISHED

Nachocss committed:
Unpublished proposals redirects to proposal listing.
Finished proposals can be fetched, but not listed

Nachocss committed:
Reviser can now list every proposal.
* Status REJECTED replaced by INADEQUATE.
* Fixed JpaStatus.findById, it now filters by natural key.

mariasoria committed:
Added method to fetch the listed volunteers in a proposal

mariasoria committed:
Added test for proposal not found and for checking the endpoint works.

mariasoria committed:
* Update Test "return_200_and_a_list_of_volunteers_from_a_proposal_as_reviser"
* Commented Role Reviser for fetchListedVolunteersInProposal ProposalApiController's method
* Created methods registerESALAndProposalWithInscribedVolunteers to create a proposal with the inscribed volunteers

mariasoria committed:
* Added JOIN FETCH to findByNaturalId to get the inscribedVolunteers

juanmaMacGyverCode committed:
Added role reviser and refactored the tests

juanmaMacGyverCode committed:
Added endpoint fetch information proposal with volunteers

pedroamador committed:
Use 0.11.5 localstack docker image version instead latest

mariasoria committed:
Added swagger comments to controllers

mariasoria committed:
Fixed SonarCloud issues. Managed NullPointerException

mariasoria committed:
Fixed SonarLint issues. Only 3 more left. Renamed Test classes.

mariasoria committed:
Fixed SonarLint issue in SecurityConfigShould. Only 2 issues more left.

mariasoria committed:
Created endpoints. First steps

mariasoria committed:
Modified endpoints. Updated parameters of all methods From EmailConfirmation to JpaCredential.

mariasoria committed:
Created EmailRecoveryPassword value object. Modified parameters of all methods From JpaCredential EmailRecoveryPassword.

mariasoria committed:
* Added token for recovering password.
* Update column sent_on

mariasoria committed:
* Modified JpaCredential entity (added columns: hash_recovery_password and created_recovery_hash_on)
* Added new version of FlyWay adding columns: hash_recovery_password and created_recovery_hash_on.
* Created query in JpaCredentialRepository (updateHashByEmail(email, hash))

mariasoria committed:
fix: new version of flyway migration.

juanmaMacGyverCode committed:
*Implemented test for endpoints send email recovery\n
*Started the implementation executePasswordChanging

juanmaMacGyverCode committed:
Security: tests recovery passwords endpoints without authentication, add endpoints in SecurityConfig

juanmaMacGyverCode committed:
Add: Manage exception in case for time has expired and add the test

mariasoria committed:
feature:
* Implemented update password in DB.
* Implemented send email confirmation after changing password.

juanmaMacGyverCode committed:
Fixed: setRecoveryPasswordHashAndDate query fix.
Test: Added test for RecoveryPasswordApiController.
Swagger: Added swagger documentation in RecoveryPasswordApiController.
Javadoc: Added Javadoc FetchCredentialsAction.

manumoranchel commented:
Not sure if it is just me of there are changes from revious Pull Request

Please, review it. Did you created the branch from develop?

Maybe you can fix it by creating a new branch and cherry-picking the new commits. And remeber always always always test :-)

mariasoria committed:
Merge branch 'mergingTest' into feature/recover-password

# Conflicts:
#	backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
#	backend/src/main/java/com/huellapositiva/application/controller/VolunteerApiController.java
#	backend/src/main/java/com/huellapositiva/application/dto/ProposalRequestDto.java
#	backend/src/main/java/com/huellapositiva/domain/actions/DeleteESALAction.java
#	backend/src/main/java/com/huellapositiva/domain/actions/EmailConfirmationAction.java
#	backend/src/main/java/com/huellapositiva/domain/actions/FetchProposalAction.java
#	backend/src/main/java/com/huellapositiva/domain/actions/RegisterESALAction.java
#	backend/src/main/java/com/huellapositiva/domain/actions/RegisterProposalAction.java
#	backend/src/main/java/com/huellapositiva/domain/actions/ResendEmailConfirmationAction.java
#	backend/src/main/java/com/huellapositiva/domain/actions/SubmitProposalRevisionAction.java
#	backend/src/main/java/com/huellapositiva/domain/actions/UploadCurriculumVitaeAction.java
#	backend/src/main/java/com/huellapositiva/domain/model/entities/Proposal.java
#	backend/src/main/java/com/huellapositiva/domain/model/entities/User.java
#	backend/src/main/java/com/huellapositiva/domain/model/valueobjects/Email.java
#	backend/src/main/java/com/huellapositiva/domain/model/valueobjects/ProposalStatus.java
#	backend/src/main/java/com/huellapositiva/domain/repository/ContactPersonRepository.java
#	backend/src/main/java/com/huellapositiva/domain/repository/CredentialsRepository.java
#	backend/src/main/java/com/huellapositiva/domain/repository/ESALContactPersonRepository.java
#	backend/src/main/java/com/huellapositiva/domain/repository/ProposalRepository.java
#	backend/src/main/java/com/huellapositiva/domain/service/ESALContactPersonService.java
#	backend/src/main/java/com/huellapositiva/domain/service/EmailCommunicationService.java
#	backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java
#	backend/src/main/java/com/huellapositiva/domain/service/RemoteStorageService.java
#	backend/src/main/java/com/huellapositiva/infrastructure/TemplateService.java
#	backend/src/main/java/com/huellapositiva/infrastructure/orm/entities/JpaCredential.java
#	backend/src/main/java/com/huellapositiva/infrastructure/orm/entities/JpaProposal.java
#	backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaCredentialRepository.java
#	backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaProposalRepository.java
#	backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql
#	backend/src/test/java/com/huellapositiva/api/ESALControllerShould.java
#	backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java
#	backend/src/test/java/com/huellapositiva/util/TestData.java

mariasoria committed:
fix: Conflicts solved

mariasoria committed:
fix: sonarLint issues solved

ibaiul commented on backend/src/main/java/com/huellapositiva/application/controller/RecoveryPasswordApiController.java:
```diff
@@ -0,0 +1,85 @@
+                    )
+            }
+    )
+    @GetMapping("/sendRecoveryPasswordEmail/{email}")
```
1- A request that causes state changes in the backed should be a POST request.

2- The email address of someone is very easily guessable so we must not use it in the request path, instead we should send it in the body.

3- This request does not require any role so it should be CSRF protected to avoid anyone making this request outside our domain

juanmaMacGyverCode replied:
Good morning!!! Yes, you are right. We fixed it.

ibaiul commented on backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java:
```diff
@@ -0,0 +1,16 @@
+ * This exception is thrown when a proposal is neither PUBLISHED nor FINISHED
+ * @see com.huellapositiva.domain.model.valueobjects.ProposalStatus
+ */
+public class ProposalNotPublic extends RuntimeException {
```
Did you meant ProposalNotPublicException?

juanmaMacGyverCode replied:
Ok, many thanks. This class has been deleted. There is other class with same exception, with "exception".

ibaiul commented on backend/src/main/java/com/huellapositiva/application/controller/RecoveryPasswordApiController.java:
```diff
@@ -0,0 +1,85 @@
+        try {
+            credentialsAction.executePasswordChanging(hash, password);
+        } catch (TimeForRecoveringPasswordExpiredException e) {
+            throw new ResponseStatusException(HttpStatus.LOCKED, "The resource is locked because time for recovery password has expired");
```
403 is a more common response

juanmaMacGyverCode replied:
Ok Ibai, fixed it.

ibaiul commented on backend/src/main/java/com/huellapositiva/application/controller/RecoveryPasswordApiController.java:
```diff
@@ -0,0 +1,85 @@
+    )
+    @PostMapping("/changePassword")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void changePassword(@RequestParam("hash") String hash, @RequestParam("newPassword") String password){
```
We should not use request parameters for this purpose. We are implementing a JSON API and we should favour this approach unless inevitable.

The exception could be the hash, that could be considered an id of a resource and therefore be part of the URI

`POST /changePassword/{hash}`
{
  "newPassword": "your-new-password"
}

juanmaMacGyverCode replied:
Ok, Ibai, we fixed it.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/actions/FetchCredentialsAction.java:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+
+    /**
+     * This method adds an hour to the date
+     *
+     * @param date
+     * @return It is the date with an hour added.
+     */
+    private Date addAnHour(Date date) {
+        Calendar c = Calendar.getInstance();
+        c.setTime(date);
+        c.add(Calendar.HOUR, 1);  // number of days to add
+        return c.getTime();
+    }
```
Java Time API has built-in functionality for this. Don't reinvent the wheel :-)

Date and Calendar objects are obsolete, favour using the new Time API whenever possible -> Instant, LocalDate, LocalDateTime, ...

juanmaMacGyverCode replied:
Ok Ibai, we have updated the Date/Calendar in Database to LocalDateTime. Many thanks, please review it.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/actions/SubmitProposalRevisionAction.java:
```diff
@@ -38,7 +38,8 @@ public void execute(String proposalId, ProposalRevisionDto revisionDto, URI prop
         Reviser reviser = Reviser.from(credentialsRepository.findByEmail(revisionDto.getReviserEmail()));
-        if (revisionDto.getHasFeedback() == true && revisionDto.getFeedback() == null) {
+        Boolean hasFeedback =  revisionDto.getHasFeedback();
+        if (hasFeedback != null && hasFeedback && revisionDto.getFeedback() == null) {
```
Why not use boolean?

juanmaMacGyverCode replied:
The sonarlint recomendation was that. What do you suggest?

ibaiul replied:
Sonar recommended to write the code in that way in case you use a Boolean object, but is not telling you whether you should use a Boolean or a boolean. That is up to your requirements.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/exception/InvalidStatusId.java:
```diff
@@ -0,0 +1,7 @@
+package com.huellapositiva.domain.exception;
+
+public class InvalidStatusId extends RuntimeException{
```
Name

juanmaMacGyverCode replied:
The class was duplicated. We removed the class.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/exception/TimeForRecoveringPasswordExpiredException.java:
```diff
@@ -0,0 +1,7 @@
+package com.huellapositiva.domain.exception;
+
+public class TimeForRecoveringPasswordExpiredException extends RuntimeException {
```
If you annotate this class with @ResponseStatus you won't need to catch this exception in the controller. Up to you

juanmaMacGyverCode replied:
We changed it, but we have some doubts about which way is more adequate?

ibaiul replied:
It is up to you how much Spring "magic" you want to use in your code. Under a design perspective you should not use the magic since if couples your code to the concrete framework that is performing that magic, makes units less testable and hides part of the logic to mention a few drawbacks.

There are lots of cons/pros around the topic. My personal taste is to use the magic as long as I know what that magic does behind the scenes and does not influence in my testing strategy. That allows a very fast development lifecycle once you know what you are doing,  but would not recommend it to someone that is getting to know Spring framework and overall application development.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/model/valueobjects/Email.java:
```diff
@@ -41,6 +41,25 @@ public static Email createFrom(ProposalRevisionEmail proposalRevisionEmail, Emai
                 .body(emailTemplate.getParsedTemplate())
                 .build();
     }
+    public static Email createFrom(EmailRecoveryPassword emailRecoveryPassword, EmailTemplate emailTemplate, String from) {
+        validateEmail(emailRecoveryPassword.getEmail());
+        return Email.builder()
+                .from(from)
+                .to(emailRecoveryPassword.getEmail())
+                .subject("Cambio de tu contraseña")
+                .body(emailTemplate.getParsedTemplate())
+                .build();
+    }
+
+    public static Email createFrom(EmailAddress emailAddress, EmailTemplate emailTemplate, String from) {
+        validateEmail(emailAddress.toString());
+        return Email.builder()
+                .from(from)
+                .to(emailAddress.toString())
+                .subject("Confirmacion de cambio de contraseña")
+                .body(emailTemplate.getParsedTemplate())
+                .build();
+    }
```
If you would pass the subject as an argument or include it in the EmailTemplate object, you would get rid of all these extra methods.

Naming "createFrom" to all methods is misleading when you have clearly different implementations. You could name the methods after their intention, e.g. createPasswordConfirmationEmailFrom, createPasswordChangeEmailFrom, ...

juanmaMacGyverCode replied:
Mira Ibai, estamos de acuerdo contigo con que el código se repite, pero no solo se repite en esto dos métodos, se repiten en 5 exactamente pero con pequeñas modificaciones. Si te parece bien, consideramos que lo mejor es verlo en el refactor general después de esta pull request porque se tocan cosas más allá del contexto de esta rama.

ibaiul replied:
You can decide when you tackle the issue, that is fine for me, just pointing out.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/repository/ESALRepository.java:
```diff
@@ -68,6 +68,7 @@ public String saveAsReviser(ESAL model) {
         }
     }
 
+
```
One line was not enough?

As an observation, when you send a PR for review you should check first all the changes you made one by one to ensure you don't include changes like this which increase the amount of changed files to review.

juanmaMacGyverCode replied:
You have right!!!!

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/service/EmailCommunicationService.java:
```diff
@@ -38,7 +38,7 @@ public void sendRegistrationConfirmationEmail(EmailConfirmation emailConfirmatio
             emailService.sendEmail(email);
         } catch (RuntimeException ex) {
-            log.error("Failed to send email:", ex);
+            log.error("Failed to send emairegisterVolunteerIssuel:", ex);
```
:-)

juanmaMacGyverCode replied:
Sorry Ibai, was a misspelling.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/service/EmailCommunicationService.java:
```diff
@@ -69,4 +69,26 @@ public void sendSubmittedProposalRevision(ProposalRevisionEmail proposalRevision
+            EmailTemplate emailTemplate = templateService.getRecoveryEmailTemplate(emailRecoveryPassword.getHash());
+            Email email = Email.createFrom(emailRecoveryPassword, emailTemplate, from);
+            emailService.sendEmail(email);
+        } catch (RuntimeException ex) {
```
Do not throw or catch RuntimeExceptions unless necessary. Create more concrete exception for your use case.

If we are not able to send the email, shouldn't we respond with 500? Otherwise the user will think the process went fine and will expect to get an email in its inbox. Which will not finally arrive and will make it get pissed.

ibaiul replied:
Create API tests for this case please if you haven't already. It is an unhappy but necessary path

juanmaMacGyverCode replied:
Mira Ibai, nos hemos estado fijando con todo lo que nos has dicho y hay muchas cosas que o están repetidas (pero usadas en otros métodos) o incluso hay que tocar muchos tests y clases que no son de esta rama. Yo creo que si hacemos ese refactor ahora, lo demás quedaría raro o bien nos estariamos excediendo al propósito de esta rama. Lo tendremos en cuenta este mensaje para la siguiente rama que haremos que es un refactor general.

Saludos.

ibaiul replied:
ok

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/service/EmailCommunicationService.java:
```diff
@@ -69,4 +69,26 @@ public void sendSubmittedProposalRevision(ProposalRevisionEmail proposalRevision
+            Email email = Email.createFrom(emailAddress, emailTemplate, from);
+            emailService.sendEmail(email);
+        } catch (RuntimeException ex) {
+            log.error("Failed to send email:", ex);
```
Do not throw or catch RuntimeExceptions unless necessary. Create more concrete exception for your use case.

If we are not able to send the email, shouldn't we respond with 500? Otherwise the user will think the process went fine and will expect to get an email in its inbox. Which will not finally arrive and will make it get pissed.

ibaiul replied:
Create API tests for this case please if you haven't already. It is an unhappy but necessary path

juanmaMacGyverCode replied:
En el mensaje de arriba pusimos eso, que lo veremos mejor en el refactor general en otra rama que no sea esta porque excede el propósito de esta rama.

ibaiul commented on backend/src/main/java/com/huellapositiva/infrastructure/orm/entities/EmailRecoveryPassword.java:
```diff
@@ -0,0 +1,39 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class EmailRecoveryPassword implements Serializable{
```
Serializable {

I have seen it many times in this PR. Missing space before curly brace as per the naming convention.

juanmaMacGyverCode replied:
Fixed it.

ibaiul commented on backend/src/main/java/com/huellapositiva/infrastructure/orm/entities/EmailRecoveryPassword.java:
```diff
@@ -0,0 +1,39 @@
+        @Column(name = "sent_on")
+        private Date sentOn;
+
+        /// ¿¿??
```
??

juanmaMacGyverCode replied:
Fixid it.

ibaiul commented on backend/src/main/java/com/huellapositiva/infrastructure/orm/entities/JpaCredential.java:
```diff
@@ -46,4 +48,12 @@
+    @Column(name = "hash_recovery_password", unique = true)
+    private String hashRecoveryPassword;
+
+    //@CreationTimestamp
```
Why not delegate the creation to the DB engine?

juanmaMacGyverCode replied:
Fixed it.

ibaiul commented on backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaStatusRepository.java:
```diff
@@ -0,0 +1,17 @@
+package com.huellapositiva.infrastructure.orm.repository;
+
+import com.huellapositiva.infrastructure.orm.entities.JpaStatus;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+
+import java.util.Optional;
+
+@Repository
+public interface JpaStatusRepository extends JpaRepository<JpaStatus, Integer> {
+
+    @Query("FROM JpaStatus s WHERE s.id = :id")
+    Optional<JpaStatus> findById(@Param("id") Integer id);
+
+}
```
What is this class for? What is JpaStatus, is it JpaProposalStatus? It has a very ambiguous name.

Actually it is not used.

juanmaMacGyverCode replied:
Ok Ibai, is fixed the problem. Delete JpaStatus.

ibaiul commented on backend/src/main/java/com/huellapositiva/infrastructure/orm/entities/JpaStatus.java:
```diff
@@ -0,0 +1,30 @@
+package com.huellapositiva.infrastructure.orm.entities;
+
+import lombok.*;
+import org.hibernate.annotations.NaturalId;
+
+import javax.persistence.*;
+import java.io.Serializable;
+
+@Entity
+@Table(name = "statuses")
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode
+@ToString
+public class JpaStatus implements Serializable {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Integer surrogateKey;
+
+    @NaturalId
+    @Column(name = "id")
+    private Integer id;
+
+    @Column(name = "name", nullable = false, unique = true)
+    private String name;
+
+}
```
This class does not seem to be used

juanmaMacGyverCode replied:
The class has been deleted.

juanmaMacGyverCode committed:
Fixed: Updated http method post in controller.
Fixed: Removed ProposalNotPublic, was duplicated.
Fixed: Changed response status http LOCKED to FORBIDDEN
Fixed: Added hash as endpoint.
Fixed: Changed calendar/date to LocalDateTime.

juanmaMacGyverCode committed:
Fixed: Remove duplicated exception.

juanmaMacGyverCode committed:
Fixed: Treatment the exception in local.

juanmaMacGyverCode committed:
Fixed: Misspelling display message error.

juanmaMacGyverCode committed:
Add space between Serializable and {.

juanmaMacGyverCode committed:
Fixed: Delete JpaStatus because is unused.

mariasoria committed:
Merge branch 'develop' into feature/recover-password
